pkg/controller/controller.go

```diff
@@ -157,17 +169,30 @@ func toSync(resyncPeriod time.Duration) entitystore.Filter {
 	}
 }
 
+// DefaultSync simply returns a list of entities in non-READY state which have been modified since the resync period.
+func DefaultSync(store entitystore.EntityStore, entityType reflect.Type, organizationID string, resyncPeriod time.Duration) ([]entitystore.Entity, error) {
+	valuesPtr := reflect.New(reflect.SliceOf(entityType))
+	if err := store.List(organizationID, defaultSyncFilter(resyncPeriod), valuesPtr.Interface()); err != nil {
+		return nil, err
+	}
+	values := valuesPtr.Elem()
+	var entities []entitystore.Entity
+	for i := 0; i < values.Len(); i++ {
+		e := values.Index(i).Interface().(entitystore.Entity)
+		entities = append(entities, e)
+	}
+	return entities, nil
+}
+
 func (dc *DefaultController) sync() error {
 	defer trace.Trace("")()
 
-	for entityType := range dc.entityHandlers {
-		entitiesPtr := reflect.New(reflect.SliceOf(entityType))
-		if err := dc.store.List(dc.options.OrganizationID, toSync(dc.options.ResyncPeriod), entitiesPtr.Interface()); err != nil {
+	for _, handler := range dc.entityHandlers {
+		entities, err := handler.Sync(dc.options.OrganizationID, dc.options.ResyncPeriod)
```

Comment:
I think this is a very good refactor. 

As the real intention of a Sync function is to find the difference between the desired status (in entity store) and the actual status (in underlying driver); then try to resolve the difference. 

I gave up and decide to work on the easier implementation first, i.e. just enforce the actual status towards the desired status. In which case, there may be some zombie entities/resources in the underlying drivers.  

I think the new Sync function of Image Manager did a good job on this, (i.e. also pull image status in the driver (image registry). Hopefully, we can apply a similar pattern for other entities. (however, sometimes it may be error-prone, difficult or inefficient to periodically "deeply" compare two set of objects/status

Reply:
👍 

LOL "zombie entities" 😂